### ISSUE TITLE
TypeError with ForceBalance-1.9.4

### ISSUE THREAD

**Issue by rvkrishnan30** [leeping/forcebalance#280]
Hi, 

I am running forcebalance-1.9.4 on a small test molecule and it fails with a TypeError. Any suggestions to fix this would be really helpful. 

Error message (optimize.err):
```
Warning: importing 'simtk.openmm' is deprecated.  Import 'openmm' instead.
Traceback (most recent call last):
  File "/home/venkat/miniconda3/envs/test_jan23/bin/ForceBalance.py", line 41, in Run_ForceBalance
    objective   = Objective(options, tgt_opts, forcefield)
  File "/home/venkat/miniconda3/envs/test_jan23/lib/python3.9/site-packages/forcebalance/objective.py", line 182, in __init__
    else: Tgt = Implemented_Targets[opts['type']](options,opts,forcefield)
  File "/home/venkat/miniconda3/envs/test_jan23/lib/python3.9/site-packages/forcebalance/smirnoffio.py", line 770, in __init__
    super(TorsionProfileTarget_SMIRNOFF,self).__init__(options,tgt_opts,forcefield)
  File "/home/venkat/miniconda3/envs/test_jan23/lib/python3.9/site-packages/forcebalance/torsion_profile.py", line 58, in __init__
    self.engine = self.engine_(target=self, mol=self.mol, **engine_args)
  File "/home/venkat/miniconda3/envs/test_jan23/lib/python3.9/site-packages/forcebalance/smirnoffio.py", line 265, in __init__
    super(SMIRNOFF,self).__init__(name=name, **kwargs)
  File "/home/venkat/miniconda3/envs/test_jan23/lib/python3.9/site-packages/forcebalance/openmmio.py", line 626, in __init__
    super(OpenMM,self).__init__(name=name, **kwargs)
  File "/home/venkat/miniconda3/envs/test_jan23/lib/python3.9/site-packages/forcebalance/engine.py", line 80, in __init__
    self.prepare(**kwargs)
  File "/home/venkat/miniconda3/envs/test_jan23/lib/python3.9/site-packages/forcebalance/smirnoffio.py", line 443, in prepare
    self.xyz_omms.append((positions, box))
  File "/home/venkat/miniconda3/envs/test_jan23/lib/python3.9/site-packages/openmm/unit/quantity.py", line 773, in append
    raise TypeError("Cannot append item without units into list with units")
TypeError: Cannot append item without units into list with units
```

**Additional information to help reproduce this:**
I created a new conda env with the following commands
```
conda create -n test_jan23 -y -c conda-forge mamba python=3.9
conda activate test_jan23
mamba install -c conda-forge openff-toolkit=0.11.4
mamba install -c conda-forge forcebalance
```
then ran ForceBalance with `ForceBalance optimize.in`

This compressed folder has all required files: [test_mol_FB_error.zip](https://github.com/leeping/forcebalance/files/10457217/test_mol_FB_error.zip)

Please let me know if you need any additional info on this. 

Thank you
Venkat

**Comment by rvkrishnan30** [comment]
Thanks @j-wags for looking into this. I checked the [code changes](https://github.com/leeping/forcebalance/pull/278/files) and copied them to my local version to test. This time I get a `FloatingPointError`

```
Warning: importing 'simtk.openmm' is deprecated.  Import 'openmm' instead.
Traceback (most recent call last):
  File "/home/venkat/miniconda3/envs/test_jan23/bin/ForceBalance.py", line 45, in Run_ForceBalance
    optimizer.Run()
  File "/home/venkat/miniconda3/envs/test_jan23/lib/python3.9/site-packages/forcebalance/optimizer.py", line 322, in Run
    xk = self.OptTab[self.jobtype]()
  File "/home/venkat/miniconda3/envs/test_jan23/lib/python3.9/site-packages/forcebalance/optimizer.py", line 950, in NewtonRaphson
    return self.MainOptimizer(b_BFGS=0)
  File "/home/venkat/miniconda3/envs/test_jan23/lib/python3.9/site-packages/forcebalance/optimizer.py", line 496, in MainOptimizer
    data        = self.Objective.Full(xk,Ord,verbose=True)
  File "/home/venkat/miniconda3/envs/test_jan23/lib/python3.9/site-packages/forcebalance/objective.py", line 314, in Full
    Objective = self.Target_Terms(vals, Order, verbose, customdir)
  File "/home/venkat/miniconda3/envs/test_jan23/lib/python3.9/site-packages/forcebalance/objective.py", line 255, in Target_Terms
    Ans = Funcs[Order](mvals, customdir=customdir)
  File "/home/venkat/miniconda3/envs/test_jan23/lib/python3.9/site-packages/forcebalance/target.py", line 319, in get_H
    Ans = self.meta_get(mvals,1,1,customdir=customdir)
  File "/home/venkat/miniconda3/envs/test_jan23/lib/python3.9/site-packages/forcebalance/target.py", line 584, in meta_get
    Answer = self.get(mvals, AGrad, AHess)
  File "/home/venkat/miniconda3/envs/test_jan23/lib/python3.9/site-packages/forcebalance/torsion_profile.py", line 186, in get
    dV[p,:], _ = f12d3p(fdwrap(compute, mvals, p), h = self.h, f0 = V)
  File "/home/venkat/miniconda3/envs/test_jan23/lib/python3.9/site-packages/forcebalance/finite_difference.py", line 110, in f12d3p
    fp = (f1-fm1)/(2*h)
FloatingPointError: invalid value encountered in divide
```

As you would notice from the optimize.out in the attached folder, all the gradients are zero. [test_mol_FB_error_FloatingPointError.zip](https://github.com/leeping/forcebalance/files/10466300/test_mol_FB_error_FloatingPointError.zip)

Given this example worked with forcebalance-1.9.3, I played around with `smirnoffio.py` by commenting off lines using `interchange` and testing `openmm_topology` generated by openff-toolkit's `to_openmm()` and also with `_openff_to_openmm_topology `as in Forcebalance 1.9.3 but neither helped. Any suggestions on where this could be coming from?

Thank you 
Venkat

**Comment by j-wags** [comment]
Thanks @rvkrishnan30 - I also get the FloatingPointError but it's good to confirm. I just went very deep down the wrong rabbit hole with the evaluator interface before I came to my senses. I appreciate your notes on this, will keep working on getting #278 in working order.

**Comment by j-wags** [comment]
Stopping for the day, but my summary notes are:

* The repro example runs completes with the "old env" (FB 1.9.3, OFFTK 0.10.6) but gets the floating point error in "new env" (FB 1.9.4, OFFTK 0.12)
* Both envs produce identical forcebalance runs up until they report the MM energy. So the input config files are parsed identically, and the parameter usage stats are identical (two assignments of each of the four custom torsions)
* The runs first diverge when the MM energies are reported:

old env:
```
#==================================================================================================================================================#
#|                                Torsion Profile: test_mol, Objective =  1.22463e+00, Units = kcal/mol, Angstrom                                 |#
#|  System                                             Min(QM)    Min(MM)      Range(QM)          Range(MM)    Max-RMSD   Ene-RMSE    Obj-Fn      |#
#==================================================================================================================================================#
test_mol                                                 -180         15     0.000 - 3.208    -0.148 - 2.297      0.257     1.1066    1.2246 
----------------------------------------------------------------------------------------------------------------------------------------------------
```

new env
```
#==================================================================================================================================================#
#|                                Torsion Profile: test_mol, Objective =  1.17156e+00, Units = kcal/mol, Angstrom                                 |#
#|  System                                             Min(QM)    Min(MM)      Range(QM)          Range(MM)    Max-RMSD   Ene-RMSE    Obj-Fn      |#
#==================================================================================================================================================#
test_mol                                                 -180         15     0.000 - 3.208    -0.101 - 2.317      0.245     1.0824    1.1716 
----------------------------------------------------------------------------------------------------------------------------------------------------
```

From this, I suspect the issue may be something with:
* molecule representation 
* SMIRKS matching/parameter assignment
* atom indexing
* geometry parsing/indexing
* conformer indexing

I've been studying [the comparison between the FB 1.9.3 code and the code in my PR](https://github.com/leeping/forcebalance/compare/v1.9.3...fix-torsionprofile) a lot but nothing there seems substantially different. So my top suspect for the root cause currently is some subtle behavior change in the OpenFF stack.

**Comment by rvkrishnan30** [comment]
@j-wags , thanks for the update. 
Trying to narrow down further, another test I did was, running FB1.9.3 but this time with the latest openff-toolkit and I ended up with a `packaging.version.InvalidVersion`

```
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/forcebalance/objective.py", line 182, in __init__
    else: Tgt = Implemented_Targets[opts['type']](options,opts,forcefield)
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/forcebalance/smirnoffio.py", line 843, in __init__
    super(TorsionProfileTarget_SMIRNOFF,self).__init__(options,tgt_opts,forcefield)
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/forcebalance/torsion_profile.py", line 58, in __init__
    self.engine = self.engine_(target=self, mol=self.mol, **engine_args)
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/forcebalance/smirnoffio.py", line 300, in __init__
    super(SMIRNOFF,self).__init__(name=name, **kwargs)
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/forcebalance/openmmio.py", line 626, in __init__
    super(OpenMM,self).__init__(name=name, **kwargs)
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/forcebalance/engine.py", line 80, in __init__
    self.prepare(**kwargs)
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/forcebalance/smirnoffio.py", line 431, in prepare
    self.off_topology = OffTopology.from_openmm(self.pdb.topology, unique_molecules=openff_mols)
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/openff/utilities/utilities.py", line 77, in wrapper
    return function(*args, **kwargs)
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/openff/toolkit/topology/topology.py", line 1466, in from_openmm
    remapped_mol.add_default_hierarchy_schemes()
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/openff/toolkit/topology/molecule.py", line 1340, in add_default_hierarchy_schemes
    self._add_chain_hierarchy_scheme(overwrite_existing=overwrite_existing)
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/openff/toolkit/topology/molecule.py", line 1349, in _add_chain_hierarchy_scheme
    self.add_hierarchy_scheme(("chain_id",), "chains")
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/openff/toolkit/topology/molecule.py", line 1420, in add_hierarchy_scheme
    self.update_hierarchy_schemes([iterator_name])
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/openff/toolkit/topology/molecule.py", line 1495, in update_hierarchy_schemes
    hierarchy_scheme.perceive_hierarchy()
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/openff/toolkit/topology/molecule.py", line 5659, in perceive_hierarchy
    self.sort_hierarchy_elements()
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/openff/toolkit/topology/molecule.py", line 5689, in sort_hierarchy_elements
    self.hierarchy_elements.sort(key=sort_func)
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/openff/toolkit/topology/molecule.py", line 5687, in sort_func
    return version.parse(".".join([str(i) for i in x.identifier]))
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/packaging/version.py", line 52, in parse
    return Version(version)
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/packaging/version.py", line 197, in __init__
    raise InvalidVersion(f"Invalid version: '{version}'")
packaging.version.InvalidVersion: Invalid version: ' '

```

I am sure you've seen this error pop up too and I am now trying to compare what changed in the toolkit that is causing this.

**Comment by j-wags** [comment]
Ah - right - you'll want to use `packaging` version 21. We're working on fixing a break due to a behavior change in packaging 22.

**Comment by rvkrishnan30** [comment]
Ah ok. Thank you for letting me know.
Also tested old openff toolkit with FB 1.9.4 and it didn't complain (of
course suitably editing little and probably that made it very similar to
1.9.3).
It seems something is misbehaving in the openff stack like you suggested.
I'll now start digging into the updates that went into openff toolkit to
see if I can get some idea.

Regards
Venkat

**Comment by rvkrishnan30** [comment]
hi @j-wags, 
Update: Continuing with the quest to find the source of zero grads, I installed `packaging version 21` and it stopped giving me the `InvalidVersion` error. Now I get `numpy.linalg.LinAlgError` and this appears to be a little more specific:
```
/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/forcebalance/finite_difference.py:110: RuntimeWarning: invalid value encountered in divide
  fp = (f1-fm1)/(2*h)
/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/forcebalance/finite_difference.py:111: RuntimeWarning: invalid value encountered in divide
  fpp = (fm1-2*f0+f1)/(h*h)
/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/forcebalance/optimizer.py:506: RuntimeWarning: invalid value encountered in double_scalars
  Quality = dX_actual / dX_expect
Traceback (most recent call last):
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/forcebalance/optimizer.py", line 322, in Run
    xk = self.OptTab[self.jobtype]()
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/forcebalance/optimizer.py", line 950, in NewtonRaphson
    return self.MainOptimizer(b_BFGS=0)
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/forcebalance/optimizer.py", line 676, in MainOptimizer
    dx, dX_expect, bump = self.step(xk, data, trust)
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/forcebalance/optimizer.py", line 747, in step
    Eig = np.linalg.eigh(H1)[0]            # Diagonalize Hessian
  File "<__array_function__ internals>", line 180, in eigh
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/numpy/linalg/linalg.py", line 1465, in eigh
    w, vt = gufunc(a, signature=signature, extobj=extobj)
  File "/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/numpy/linalg/linalg.py", line 95, in _raise_linalgerror_eigenvalues_nonconvergence
    raise LinAlgError("Eigenvalues did not converge")
numpy.linalg.LinAlgError: Eigenvalues did not converge
```

Dissecting FB code, narrowed this down to line **`data        = self.Objective.Full(xk,Ord,verbose=True)`** in `optimize.py` which is returning grads zero and then nan. Printing `data` at iteration 0 and 1 gives this below output before throwing `numpy.linalg.LinAlgError` error:
```
Iteration 0: 
{'X': 0.06591127567437205, 'G': array([0., 0., 0., 0.]), 'H': array([[1000001.,       0.,       0.,       0.],
       [      0., 1000001.,       0.,       0.],
       [      0.,       0., 1000001.,       0.],
       [      0.,       0.,       0., 1000001.]]), 'X0': 0.06591127567437205, 'G0': array([0., 0., 0., 0.]), 'H0': array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])}
Iteration 1:
/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/forcebalance/finite_difference.py:110: RuntimeWarning: invalid value encountered in divide
  fp = (f1-fm1)/(2*h)
/home/venkat/xxx/third-party/Python/lib/python3.9/site-packages/forcebalance/finite_difference.py:111: RuntimeWarning: invalid value encountered in divide
  fpp = (fm1-2*f0+f1)/(h*h)
{'X': 0.06591127567437205, 'G': array([nan, nan, nan, nan]), 'H': array([[nan, nan, nan, nan],
       [nan, nan, nan, nan],
       [nan, nan, nan, nan],
       [nan, nan, nan, nan]]), 'X0': 0.06591127567437205, 'G0': array([nan, nan, nan, nan]), 'H0': array([[nan, nan, nan, nan],
       [nan, nan, nan, nan],
       [nan, nan, nan, nan],
       [nan, nan, nan, nan]])}
```

For reference, the working older version (old OOFTK and FB) has non-zero numbers in data:
```
{'X': 0.06591118270470155, 'G': array([-2.09013398, -0.55849699,  0.05372039,  0.00651276]), 'H': array([[ 1.00003314e+06,  8.85535261e+00, -8.51773631e-01,
        -1.03264263e-01],
       [ 8.85535261e+00,  1.00000237e+06, -2.27599288e-01,
        -2.75928625e-02],
       [-8.51773631e-01, -2.27599288e-01,  1.00000002e+06,
         2.65408660e-03],
       [-1.03264263e-01, -2.75928625e-02,  2.65408660e-03,
         1.00000000e+06]]), 'X0': 0.06591118270470155, 'G0': array([-2.09013398, -0.55849699,  0.05372039,  0.00651276]), 'H0': array([[ 3.31405072e+01,  8.85535261e+00, -8.51773631e-01,
        -1.03264263e-01],
       [ 8.85535261e+00,  2.36620608e+00, -2.27599288e-01,
        -2.75928625e-02],
       [-8.51773631e-01, -2.27599288e-01,  2.18921911e-02,
         2.65408660e-03],
       [-1.03264263e-01, -2.75928625e-02,  2.65408660e-03,
         3.21766589e-04]])}
```

Do you have any thoughts/suggestions on how to fix this when the objective function is evaluated, please?

Thank you 
Venkat

**Comment by j-wags** [comment]
Thanks for the update. I'm still working on this but haven't found anything worth reporting. 

Not urgent - what change did you make to go from the `FloatingPointError` to `numpy.linalg.LinAlgError`? I'm still getting `FloatingPointError`. 

Regardless of the final error message, I think that the 0 gradient is the right thing to be looking at - I'm gonna keep trying to figure out why that's started happening.

**Comment by j-wags** [comment]
All optimizations run by FB end in energy=26.178, which I take to mean that the parameters aren't being updated. Tomorrow I'll start by verifying that the expected physical values actually appear both in the FF and in the OpenMM system.

**Comment by rvkrishnan30** [comment]
> Thanks for the update. I'm still working on this but haven't found anything worth reporting.
> 
> Not urgent - what change did you make to go from the `FloatingPointError` to `numpy.linalg.LinAlgError`? I'm still getting `FloatingPointError`.

@j-wags : The only change I did was to downgrade to packaging 21 but still stick to the new OFFTK-0.11 and FB-1.9.4
> 
> Regardless of the final error message, I think that the 0 gradient is the right thing to be looking at - I'm gonna keep trying to figure out why that's started happening.

I am still trying to traceback which specific line is returning these zeros array for `G `and `H` hoping that can point us to the source of error. Now looking at `target.py` and `finite_difference.py` but still not completely sure how much this would help if FB1.9.4 hasn't changed much compared to the previous working implementation. 

> All optimizations run by FB end in energy=26.178, which I take to mean that the parameters aren't being updated. Tomorrow I'll start by verifying that the expected physical values actually appear both in the FF and in the OpenMM system.
Will keep fingers crossed. 

Thank you 
venkat

**Comment by mattwthompson** [comment]
Here's another test case that reproduces the `FloatingPointError: invalid value encountered in divide` above. It runs pretty quickly (< 10 seconds) since it's only an ethanol molecule. 

My local environment is version v1.9.4 with 9cae3e7df938e954f4372acae4cd9c8d7bfff5ca cherry-picked.


[issue-280.tar.gz](https://github.com/leeping/forcebalance/files/10512267/issue-280.tar.gz)

**Comment by j-wags** [comment]
I've got @rvkrishnan30's repro example not crashing using the changes from [this commit](https://github.com/leeping/forcebalance/pull/278/commits/d3397f0d7cd0eda20acdc2a27eb9c731b5ffd5c3). It looks like something in the parameter-updating machinery broke.

**Comment by rvkrishnan30** [comment]
Thank you @j-wags . With the changes in your commit it stops throwing error. 

So, my understanding so far from this [code](https://github.com/leeping/forcebalance/pull/278/commits/d3397f0d7cd0eda20acdc2a27eb9c731b5ffd5c3#diff-e94d3410bc7aa592edcea50921ab888ab787ef5e8d782aa42f03ec5f8c24c237R214) is that with `units` associated to `param_quantity` (e.g. `1.532232714475 kilocalorie / mole`), the `isinstance()` never returns `true` and so the parameters don't get updated (the `setattr` never gets executed). On the flip side, we we check `param_quantity.magnitude`, the condition gets satisfied and it enters the `if` block but now the `new_value` is dimensionless and this complains (`openff.toolkit.utils.exceptions.IncompatibleUnitError: k=1.532232714475 dimensionless should have units of kilocalorie / mole`). 
So we need to reassign the original units to `new_value` and this is best done by just taking the units from  `param_quantity.units`. 

All this said, with this change below (almost your code except that `param_quantity.unit` is `param_quantity.units`) the code seems to be happy and we can have all the parts you commented back to how it was before. 

```
if isinstance(param_quantity, (float, int)):
    print(f"smirnoffio.assign_openff_parameter 100")
    #setattr(parameter_container, value_name, new_value)
    setattr(parameter_container, value_name, new_value * param_quantity.unit)
    param_quantity = None
```

```
if isinstance(param_quantity.magnitude, (float, int)): # _convert_magnitude_not_inplace
    setattr(parameter_container, value_name, new_value * param_quantity.units)
    param_quantity = None
```

Hope this helps and please let me know if this makes sense. 
Thank you
Venkat

**Comment by j-wags** [comment]
Allllright. Thanks for much for the help @rvkrishnan30. I think I fully wrapped my head around this one. 

The true bug is at the very end of the old method:

```python
    if param_quantity is not None:
        param_quantity._value = new_value
```

Previously, `param_quantity` was an `openmm.unit.Quantity` object, which stored its unitless value information in `._value`. 

With the new (>=0.11) toolkit, it's an `openff.units.Quantity` object which stores its unitless value somewhere else, and doesn't make any use of its `._value` attribute. 

But this is python, where you can set any attribute on any object, whether or not it has any meaning. So the line above would happily set a meaningless `._value` attribute on an OpenFF quantity, raising no error but leaving the underlying physical parameters unchanged. 

(I'd love to conclude that the original sin was private attribute access, but even if it were just setting a non-private `.value` we would have had the same issue since OpenFF quantities also don't do anything with that attribute)

I think we'd eventually run into problems with sticking `* param_quantity.units` on the code path you're suggesting, since this ONLY works if the FF value is a Quantity. But this isn't true of all FF values - for example torsion periodicities really are just basic `int` and 1-4 scaling factors are `float`. So we'd get a new bug the day someone tried to fit those because trying to call`float.units` would raise an error. 

I think the right solution is to not avoid using the `ff._forcebalance_assign_parameter_map` cache (as both of our proposed code changes do), but to assume that, if we ever reach the line I quoted at the beginning, that we are dealing with a OpenFF Quantity, and THEN do the  `* param_quantity.units` trick.

Thanks for bearing with me. The old PR is a bit messy so I'll open a new one with the minimal changes either today or Monday.

**Comment by rvkrishnan30** [comment]
Thank you very much @j-wags . I'll wait for your new PR with updated changes to fully understand the best/correct solution as I am afraid I did not fully understand the right solution. 

I got the part about `param_quantity` being `openff.units.Quantity` with the new OFFTK (>v0.11) and that checking `magnitude` in `if isinstance` and doing `* param_quantity.units` might eventually throw bugs in the future if we deal with other FF which are not `Quantity`. 

Really sorry but I am perhaps missing something trivial in the code change you suggest as I keep getting `numpy.linalg.LinAlgError` or it does not enter the if block to `seattr()` without the code change suggested earlier (pasted below for reference). 

```
if isinstance(param_quantity.magnitude, (float, int)):
    setattr(parameter_container, value_name, new_value * param_quantity.units)
    param_quantity = None
``` 

I guess my questions will get cleared as soon as I see you new PR. Looking forward to it. 
Thank you
Venkat
